crates/common/mqtt_channel/src/connection.rs

```diff
@@ -0,0 +1,239 @@
+use crate::{Config, ErrChannel, Message, MqttError, PubChannel, SubChannel, TopicFilter};
+use futures::channel::mpsc;
+use futures::{SinkExt, StreamExt};
+use rumqttc::{
+    AsyncClient, ConnectionError, Event, EventLoop, Incoming, Outgoing, Packet, StateError,
+};
+use std::time::Duration;
+use tokio::time::sleep;
+
+/// A connection to some MQTT server
+pub struct Connection {
+    /// The channel of the input messages received by this connection.
+    pub received: mpsc::UnboundedReceiver<Message>,
+
+    /// The channel of the output messages to be published on this connection.
+    pub published: mpsc::UnboundedSender<Message>,
+
+    /// The channel of the error messages received by this connection.
+    pub errors: mpsc::UnboundedReceiver<MqttError>,
+}
+
+impl Connection {
+    /// The stream of events received from this MQTT connection and forwarded to the client
+    pub fn sub_channel(&self) -> &impl SubChannel {
+        &self.received
+    }
+
+    /// The stream of actions sent by the client to this MQTT connection
+    pub fn pub_channel(&self) -> &impl PubChannel {
+        &self.published
+    }
+
+    /// The stream of errors received from this MQTT connection and forwarded to the client
+    pub fn err_channel(&self) -> &impl ErrChannel {
+        &self.errors
+    }
+
+    /// Establish a connection to the MQTT broker defined by the given `config`.
+    ///
+    /// This connection is associated to an MQTT session with the given `name` and `subscription`.
+    ///
+    /// Reusing the same session name on each connection allows a client
+    /// to have its subscription persisted by the broker
+    /// so messages sent while the client is disconnected are received on re-connection.
+    ///
+    /// The connection will only receive messages published on the subscription topics.
+    ///
+    /// ```no_run
+    /// # use mqtt_channel::{Config, Connection, Topic, MqttError};
+    /// # use std::convert::TryInto;
+    ///
+    /// # #[tokio::main]
+    /// # async fn connect() -> Result<Connection, MqttError> {
+    ///     // A client can subscribe to many topics
+    ///     let topics = vec![
+    ///         "/a/first/topic",
+    ///         "/a/second/topic",
+    ///         "/a/+/pattern", // one can use + pattern
+    ///         "/any/#",       // one can use # pattern
+    ///     ]
+    ///     .try_into()
+    ///     .expect("a list of topic filters");
+    ///
+    ///     Connection::connect("test", &Config::default(), topics).await
+    /// # }
+    pub async fn connect(
+        name: &str,
+        config: &Config,
+        subscription: TopicFilter,
+    ) -> Result<Connection, MqttError> {
+        let (received_sender, received_receiver) = mpsc::unbounded();
+        let (published_sender, published_receiver) = mpsc::unbounded();
+        let (error_sender, error_receiver) = mpsc::unbounded();
```

Comment:
I'm not a fan of the `Connection` having only pub interface `connect` this will lead to similar issues we had with the previous implementation `mqtt_client` where we would have to know the server details at the points of creating the struct as well as the comms channels are created here.

I would prefer if there is some kind of `new`/`try_new` function on the struct where I can pass the channels and configs and only when I'm ready call `connect` on it or not. For example this would make tests easier to perform.

something along the lines:

```suggestion
    pub async fn new(
        name: &str,
        config: &Config,
        subscription: TopicFilter,
        channel_in: Unbounded,
        channel_out: Unbounded,
        channel_err: Unbounded
    ) -> Result<Self, MqttError> {
        ...
        Ok(Self {
        received: received_receiver,
            published: published_sender,
            errors: error_receiver,
        })
```

and then to call:

```rust
let connection = Connection::new(..., ..., ..., );
connection.connect();

```

Reply:
That is a very good point. This raises some questions though.

A connection that is not open yet is somehow a configuration for a connection.

1. Do we introduce a new `ConnectionConfiguration`  struct binding the MQTT configuration with a session configuration?
      ```
      let session = ConnectionConfiguration::new(name, mqtt_config, topics);
      let connexion = session.connect().await?
      ```
2. Do we extent the `Configuration` struct with session information?
     ```
     let config = mqtt_config.with_session(name).with_subscription(topics);
     let connexion =Connection::new(config).await?
     ```
3. Do we hide the differences between open and close configuration using an internal state?
    ```
    enum Connection{
           Config { session_name, mqtt_config, subscriptions },
           Open { sub_channel, pub_channel },
    }
    ```
    
What do you think?

I have a preference for option 2 that avoids a cascade of config structs and makes clear the expectations for a connection. Notably, a `with_session(name)` can set accordingly the clean session flag to false.

Reply:
Option 2 works for me.

Reply:
Option 2 implemented by https://github.com/thin-edge/thin-edge.io/pull/575/commits/cfd8736c9c7e7406a5b1aa841f41148168c957f3